Feature/tournament management and template edit

Kneshal opened the pull request:
Добавил view для открытия/закрытия регистрации, добавил декоратор для проверки пользователя, добавил шаблон-карточку турнира и т.д.

Kneshal committed:
Start work under tourament management

Kneshal committed:
Finish working

lejbron commented on tournaments/models.py:
```diff
@@ -34,59 +37,84 @@ class Tournament(models.Model):
-
+    superviser = models.ForeignKey(
+        User,
+        verbose_name='Организатор',
+        help_text='Укажите организатора турнира',
```
Ты на чем то основывался, когда первыми ставил verbose_name и help_text? 
Я просто думал о порядке перечисления параметров и так и не понял, чем руководствоваться)

Kneshal replied:
Просто привык так делать)

Kneshal committed:
Add icons, some fix

lejbron commented on tournaments/views/status_change_forms.py:
```diff
@@ -1,8 +1,53 @@
 from tournaments.models import Tour, Tournament
 
 
+def superviser_check(func):
+    """Декоратор. Проверяет является ли юзер организатором турнира."""
+    def check_user(request, *args, **kwargs):
+        tournament = get_object_or_404(Tournament, tt_slug=kwargs.get('tt_slug'))
+        if tournament.superviser != request.user:
+            return redirect('index')
+        return func(request, *args, **kwargs)
+    return check_user
```
Стоит все таки вынести в отдельный файл, а так же использовать wraps из functools, убирает некоторые внутренние проблемы использования - получение имени и документации. 

[Статья про декоратор](https://simpleisbetterthancomplex.com/2015/12/07/working-with-django-view-decorators.html)

Статья про functuls.wraps:
[eng](https://lerner.co.il/2019/05/05/making-your-python-decorators-even-better-with-functool-wraps/)
[rus](https://www.google.ru/amp/s/webdevblog.ru/ispolzovanie-functool-wraps-v-dekoratorah-python/amp/)

В целом решение огонь, самому даже в голову не приходило такое! 🔥

Kneshal replied:
Тут кстати еще предполагалось, что если прав нет, то редиректит на страницу 403.html. Надо бы ее добавить.

lejbron replied:
Страницу добавлю в mini_release тогда, когда все причесывать буду. Задачку сделаю, спасибо, что напомнил)

lejbron replied:
с декоратором есть проблема:
нельзя вызывать для [функций изменения статуса тура](https://github.com/lejbron/arkenstone/blob/7fc4922adb92fbdd7e8b50fcfa2202a5dcc4da16/tournaments/views/status_change_forms.py#L67) - вместо tt_slug приходит tour_slug

lejbron commented on tournaments/forms.py:
```diff
@@ -1,14 +1,22 @@
     """
     Форма создания турнира.
     """
+    superviser_id = forms.ChoiceField(
+        label='Организатор',
+        choices=User.objects.filter(is_staff=True).values_list('id', 'username'),
+        )
+
```
по идее можно было просто добавить поле в список в Meta классе, а потом при создании формы во view задать значения, из которых можно выбирать.
Пример [тут](https://github.com/lejbron/arkenstone/blob/ef1e473d89cd2fb5a302db128014b112ca353a22/tournaments/views/data_input_forms.py#L72).

Kneshal replied:
Попробую, также хотел сделать, но не нашел как.

lejbron commented on tournaments/templates/tournaments_list.html:
```diff
@@ -3,34 +3,16 @@
\ No newline at end of file
+{% if user.is_staff %}
+<a class="btn btn-dark btn-lg mb-2" href="{% url 'tournament-crt' %}" role="button">
+  Открыть новый турнир
```
Думаю все таки Создать новый турнир, а не открыть.

Kneshal replied:
Поправлю

lejbron commented on tournaments/urls.py:
```diff
@@ -13,6 +13,8 @@
 urlpatterns += [
     path('<slug:tt_slug>/register/', views.register_on_tournament, name='tournament-reg'),
     path('<slug:tt_slug>/start/', views.start_tournament, name='tournament-start'),
+    path('<slug:tt_slug>/open-registration/', views.open_registration, name='tournament-open-registration'),
+    path('<slug:tt_slug>/close-registration/', views.close_registration, name='tournament-close-registration'),
```
В name можно сократить registration до reg.

lejbron commented on tournaments/views/data_input_forms.py:
```diff
@@ -16,18 +21,20 @@ def create_tournament(request):
-
     if crt_form.is_valid():
-        crt_form.save()
+        tournament = crt_form.save(commit=False)
+        superviser_id = crt_form.cleaned_data['superviser_id']
+        superviser = get_object_or_404(User, id=superviser_id)
+        tournament.superviser = superviser
+        tournament.save()
```
Если добавить поле через Meta класс модели, достаточно будет просто сохранить форму.

Kneshal committed:
decorator hotfix

Kneshal committed:
decorator, tt_form_crt, url, templates fix

Kneshal committed:
odd number of players hotfix in view

lejbron commented on tournaments/decorators.py:
```diff
@@ -0,0 +1,22 @@
+            tournament = get_object_or_404(Tournament, tours__tour_slug=kwargs.get('tour_slug'))
+        else:
+            return func(request, *args, **kwargs)
+        if tournament.superviser != request.user:
+            raise PermissionDenied
```
получается если мы придем сюда из else, то снова будет ошибка - ведь у нас tournament определяется только в if и elif. 
Думаю нужно закладываться на то, что в функциях, в которых нельзя получить турнир, этот декоратор применяться не должен. А значит пункт с else не нужен.

Kneshal replied:
Из else мы сюда не придем, там идет возврат к функции без изменений.

lejbron replied:
да, ты прав, не придем. 
Но тогда какой смысл в декораторе для такой функции? думаю лучше получать ошибки и понимать, что декоратор к функции не применим. 

т е предлагаю else все же убрать и использовать декоратор только для функций, в которых можно получить турнир и проверить орг ли пытается с ним что-то сделать.

Kneshal replied:
Наверное тогда для большей очевидности, сделать отдельные декораторы для турнира и тура.

Kneshal replied:
Может тогда в else выбрасывать какое-нибудь исключение, например ObjectDoesNotExist

Kneshal replied:
Или SuspiciousOperation

lejbron replied:
думаю надо оставить один для турнира и тура, потом при необходимости добавить туда матч. 
можно повесить исключение на else, да - но по идее лучше чтобы в этот момент все просто ломалось - и мы понимали, что применяем этот декоратор не там, где нужно

lejbron commented on tournaments/views/data_input_forms.py:
```diff
@@ -38,25 +42,22 @@ def register_on_tournament(request, tt_slug):
-    if reg_form.is_valid():
-        player_stat = reg_form.save(commit=False)
-        player_stat.tournament = tournament
-        player_stat.player = request.user
-        player_stat.save()
-        return redirect('index')
-
     context = {
         'tournament': tournament,
         'reg_form': reg_form,
     }
-
-    return render(request, 'tournament_reg_form.html', context)
+    if not reg_form.is_valid() or tournament.superviser == request.user:
+        return render(request, 'tournament_reg_form.html', context)
+    player_stat = reg_form.save(commit=False)
+    player_stat.tournament = tournament
+    player_stat.player = request.user
+    player_stat.save()
+    return redirect('index')
```
а зачем в рамках этой задачи трогать регистрацию? 
конструкция с if not очень непонятно читается.

Kneshal replied:
Это мне ревьюер так просил делать в рамках практикума:) Поменять?

lejbron replied:
нет, я понял почему так - чтобы else не писать, сначала показалось странным. оставляем так!

lejbron commented on tournaments/views/status_change_forms.py:
```diff
@@ -1,8 +1,47 @@
+    tournament = get_object_or_404(Tournament, tt_slug=tt_slug)
+    count = tournament.players.count()
+    if (count % 2 == 1) or (count == 0):
+        return redirect('tournaments-list')
```
нужно возвращать на страницу турнира, а не страницу со списком

Kneshal replied:
Ок

Kneshal replied:
Поправил до более правильной конструкции:    
if (count % 2 == 0) or (count > 0):
 -----tournament.tt_status = 'creg'
 -----tournament.save()
return redirect('tournament-detail', tournament.tt_slug)

lejbron replied:
только тут должно было быть and, а не or - но я уже поменял, когда модальное окно правил

Kneshal committed:
check players num, fix part/leave btns

Kneshal committed:
decorator, close reg, tt reg fix

Commit:
Update Modal Warning
- Delete old bootstrap include in base_generic.html
- Update modal and script in tournament_card.html
- Add control flag to close_registration view

lejbron reviewed (approved):
:fire:

lejbron committed:
Merge branch 'release/mini_release_test' into feature/tournament_management_and_template_edit

Kneshal committed:
change registre on tt view - syntax
